### PR TITLE
fix(ansible/rocker): use ROS 2 apt repository for Ubuntu 22.04

### PR DESCRIPTION
**Pull request by kenji-miyake**
## Description

There isn't `jammy` in the `ros` repository, but is in the `ros2` repository.

http://packages.ros.org/ros/ubuntu/dists/
http://packages.ros.org/ros2/ubuntu/dists/

This causes an error https://github.com/autowarefoundation/autoware/discussions/239#discussioncomment-2645081.

I confirmed `python3-rocker` is listed here.
http://packages.ros.org/ros2/ubuntu/dists/jammy/main/binary-amd64/Packages

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by kenji-miyake** [comment]
Ah, I have to fix the document as well.
-> It already uses ROS 2. https://github.com/autowarefoundation/autoware/blob/2b1d5ea1e86b53bacb890e99acedca97a0af7aae/ansible/roles/rocker/README.md?plain=1#L21

File: ansible/roles/rocker/tasks/main.yaml [hunk 1]
```diff
@@ -1,14 +1,26 @@
...
+    dest: /usr/share/keyrings/ros-archive-keyring.gpg
 
-# echo "deb http://packages.ros.org/ros/ubuntu $(lsb_release -cs) main" | sudo tee /etc/apt/sources.list.d/ros-latest.list > /dev/null
-- name: Add OSRF apt repository to source list
+- name: Save result of 'dpkg --print-architecture'
+  ansible.builtin.command: dpkg --print-architecture
+  register: deb_architecture
+  changed_when: false
+
+- name: Save result of 'source /etc/os-release && echo $UBUNTU_CODENAME'
```
Comment (esteve) [inline ansible/roles/rocker/tasks/main.yaml]: It's better to use `lsb_release -cs` as it's defined by the LSB standard

**Review by esteve — CHANGES_REQUESTED** [review]

File: ansible/roles/rocker/tasks/main.yaml [hunk 2]
```diff
@@ -1,14 +1,26 @@
-- name: Authorize OSRF GPG key
+# sudo curl -sSL https://raw.githubusercontent.com/ros/rosdistro/master/ros.key -o /usr/share/keyrings/ros-archive-keyring.gpg
+- name: Authorize ROS GPG key
   become: true
-  ansible.builtin.apt_key:
-    url: https://raw.githubusercontent.com/ros/rosdistro/master/ros.asc
+  ansible.builtin.get_url:
+    url: https://raw.githubusercontent.com/ros/rosdistro/master/ros.key
+    dest: /usr/share/keyrings/ros-archive-keyring.gpg
```
Comment (esteve) [inline ansible/roles/rocker/tasks/main.yaml:6]: The keyring may contain other keys, it's better to add it via `ansible.builtin.apt_key`

**Review by esteve — CHANGES_REQUESTED** [review]

File: ansible/roles/rocker/tasks/main.yaml [hunk 3]
```diff
@@ -1,14 +1,26 @@
...
+- name: Save result of 'source /etc/os-release && echo $UBUNTU_CODENAME'
+  ansible.builtin.shell: bash -c 'source /etc/os-release && echo $UBUNTU_CODENAME'
+  register: ubuntu_codename
+  changed_when: false
+
+# echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/ros-archive-keyring.gpg] http://packages.ros.org/ros2/ubuntu $(source /etc/os-release && echo $UBUNTU_CODENAME) main" | sudo tee /etc/apt/sources.list.d/ros2.list > /dev/null
+- name: Add ROS 2 apt repository to source list
   become: true
   ansible.builtin.apt_repository:
-    repo: deb http://packages.ros.org/ros/ubuntu {{ lsb_release_cs.stdout }} main
```
Comment (esteve) [inline ansible/roles/rocker/tasks/main.yaml:10]: I think just updating the URL from `http://packages.ros.org/ros/ubuntu` to `http://packages.ros.org/ros2/ubuntu` is enough, there's no need to add these extra steps in this task.

**Review by esteve — CHANGES_REQUESTED** [review]

File: ansible/roles/rocker/tasks/main.yaml (hunk 1 above)
Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml]: Hmm? But ROS 2 replaced it with `UBUNTU_CODENAME`. 
https://github.com/ros2/ros2_documentation/pull/2104

File: ansible/roles/rocker/tasks/main.yaml (hunk 2 above)
Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml:6]: I just followed this page. Could you explain the problem a bit more in detail?
https://docs.ros.org/en/rolling/Installation/Ubuntu-Install-Binary.html#add-the-ros-2-apt-repository

File: ansible/roles/rocker/tasks/main.yaml (hunk 3 above)
Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml:10]: What are the extra steps? I believe they all are in this installation guide. :thinking: 
https://docs.ros.org/en/rolling/Installation/Ubuntu-Install-Binary.html#add-the-ros-2-apt-repository

Comment (esteve) [reply ansible/roles/rocker/tasks/main.yaml:10]: I meant in this PR, we only need to update the URL, we don't need the extra tasks (i.e. `deb_architecture`, `ubuntu_codename`). The code we had before was enough.

File: ansible/roles/rocker/tasks/main.yaml (hunk 1 above)
Comment (esteve) [reply ansible/roles/rocker/tasks/main.yaml]: But it only works with Ubuntu, whereas `lsb_release -cs` is standard for all Linux distributions, so the current code works for Debian as well.

File: ansible/roles/rocker/tasks/main.yaml (hunk 3 above)
Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml:10]: Technically, it's true.

But,
- `/usr/share/keyrings` is the recommended way to manage keys.
- We already use this way in the [ros2](https://github.com/autowarefoundation/autoware/blob/f6a2b5254050d2a055a693e075fef598ea0b744e/ansible/roles/ros2/tasks/main.yaml) role. We don't have to duplicate the key.

File: ansible/roles/rocker/tasks/main.yaml (hunk 2 above)
Comment (esteve) [reply ansible/roles/rocker/tasks/main.yaml:6]: `ansible.builtin.apt_key` correctly calls `apt-key add` to add a remote key, `get_url` replaces the file in the filesystem with a remote file. If there are any other keys in that keyring, `get_url` will overwrite them, but `apt_key` will add the new key while maintaining the existing key.

**Review by esteve — CHANGES_REQUESTED** [review]
Just updating the URL is enough, we don't need to change anything else in the `main.yaml` file

File: ansible/roles/rocker/tasks/main.yaml (hunk 1 above)
Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml]: I see. Then it's okay to use `lsb_release -cs` for the Docker-related roles.

Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml]: Replaced it in 9e914d29a5db4ba0b4b197600b06f0044e5214ed.

File: ansible/roles/rocker/tasks/main.yaml (hunk 2 above)
Comment (kenji-miyake) [reply ansible/roles/rocker/tasks/main.yaml:6]: I think it's not a recommended way and replaced in this PR. Do you really recommend using `apt_key`? :thinking: 
https://github.com/ros2/ros2_documentation/pull/1189

Comment (esteve) [reply ansible/roles/rocker/tasks/main.yaml:6]: You're right, according to https://wiki.debian.org/DebianRepository/UseThirdParty#OpenPGP_Key_distribution

> The key MUST be downloaded over a secure mechanism like HTTPS to a location only writable by root, which SHOULD be /usr/share/keyrings. The key MUST NOT be placed in /etc/apt/trusted.gpg.d or loaded by apt-key add. 

Thanks for the link :+1:

**Review by esteve — APPROVED** [review]
Thanks for addressing my feedback :+1:

**Pull request merged**